[beta] backports

cuviper opened the pull request:
- Do not eat nested expressions' results in `MayContainYieldPoint` format args visitor #122680
- Fix heading anchors in doc pages. #122693
- Make `#[diagnostic::on_unimplemented]` format string parsing more robust #122402
-  Update ninja on Windows #123178 

r? cuviper

weiznich committed:
Make `#[diagnostic::on_unimplemented]` format string parsing more robust

This commit fixes several issues with the format string parsing of the
`#[diagnostic::on_unimplemented]` attribute that were pointed out by
@ehuss.
In detail it fixes:

* Appearing format specifiers (display, etc). For these we generate a
warning that the specifier is unsupported. Otherwise we ignore them
* Positional arguments. For these we generate a warning that positional
arguments are unsupported in that location and replace them with the
format string equivalent (so `{}` or `{n}` where n is the index of the
positional argument)
* Broken format strings with enclosed }. For these we generate a warning
about the broken format string and set the emitted message literally to
the provided unformatted string
* Unknown format specifiers. For these we generate an additional warning
about the unknown specifier. Otherwise we emit the literal string as
message.

This essentially makes those strings behave like `format!` with the
minor difference that we do not generate hard errors but only warnings.
After that we continue trying to do something unsuprising (mostly either
ignoring the broken parts or falling back to just giving back the
literal string as provided).

Fix #122391

(cherry picked from commit 5568c569c03c8c22ce81185b9e49efcaa6866050)

lqd committed:
do not eat nested exprs result in format args visitor

(cherry picked from commit 8beec62315538da7449fe869fd366181f7923b6e)

lqd committed:
add non-regression test for issue 122674

(cherry picked from commit f3e9dfaed6c4d44fc0a5182221c31e5b0ff038fd)

ehuss committed:
Fix heading anchors in doc pages.

(cherry picked from commit 39f2d25090e4b61a695fb6d72113689dcf2ae724)

cuviper commented:
@bors r+ rollup=never

bors commented:
:pushpin: Commit 052be5c643e0020eb52260fec808826aba538457 has been approved by `cuviper`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cuviper 052be5c643e0020eb52260fec808826aba538457 -->
<!-- homu: {"type":"Approved","sha":"052be5c643e0020eb52260fec808826aba538457","approver":"cuviper","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 052be5c643e0020eb52260fec808826aba538457 with merge c21a2a1280d19e5fde71bfc0417243004fe4ddba...
<!-- homu: {"type":"BuildStarted","head_sha":"052be5c643e0020eb52260fec808826aba538457","merge_sha":"c21a2a1280d19e5fde71bfc0417243004fe4ddba"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8473462168/job/23218366182)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/8473462168/job/23218366182","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`dist-aarch64-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/8473462168/job/23217838956) [(plain)](https://github.com/rust-lang-ci/rust/commit/c21a2a1280d19e5fde71bfc0417243004fe4ddba/checks/23217838956/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
-- Configuring done (33.5s)
-- Generating done (2.6s)
-- Build files have been written to: C:/a/rust/rust/build/x86_64-pc-windows-msvc/llvm/build
running: "cmake" "--build" "." "--target" "install" "--config" "Release" "--" "-j" "8"
ninja: error: build.ninja:3780: multiple outputs aren't (yet?) supported by depslog; bring this up on the mailing list if it affects you
thread 'main' panicked at C:\Users\runneradmin\.cargo\registry\src\index.crates.io-6f17d22bba15001f\cmake-0.1.48\src\lib.rs:975:5:

command did not execute successfully, got: exit code: 1

```

</details>

Nilstrieb commented:
@bors retry https://github.com/rust-lang/rust/pull/122671#issuecomment-2026178791

Noratrieb committed:
Update ninja on Windows

Errors started showing up, and I read somewhere that this might be
because of old ninja versions. This ninja version is indeed *ancient*.

```
multiple outputs aren't (yet?) supported by depslog; bring this up on the mailing list if it affects you
```

(cherry picked from commit b546764cbef7ac449cbad86636b4ca8daa9f5ef6)

cuviper commented:
@bors r+

bors commented:
:pushpin: Commit 10e6ec5ffc2a0ee11d7bf0d67509cb4354ae2040 has been approved by `cuviper`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cuviper 10e6ec5ffc2a0ee11d7bf0d67509cb4354ae2040 -->
<!-- homu: {"type":"Approved","sha":"10e6ec5ffc2a0ee11d7bf0d67509cb4354ae2040","approver":"cuviper","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 10e6ec5ffc2a0ee11d7bf0d67509cb4354ae2040 with merge c1195518c8ad7f37d15ecb82f39dec3da97f9c2f...
<!-- homu: {"type":"BuildStarted","head_sha":"10e6ec5ffc2a0ee11d7bf0d67509cb4354ae2040","merge_sha":"c1195518c8ad7f37d15ecb82f39dec3da97f9c2f"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8477186215/job/23229779144)
Approved by: cuviper
Pushing c1195518c8ad7f37d15ecb82f39dec3da97f9c2f to beta...
<!-- homu: {"type":"BuildCompleted","approved_by":"cuviper","base_ref":"beta","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8477186215/job/23229779144"},"merge_sha":"c1195518c8ad7f37d15ecb82f39dec3da97f9c2f"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8477186215/job/23229779144)
Approved by: cuviper
Pushing c1195518c8ad7f37d15ecb82f39dec3da97f9c2f to beta...
<!-- homu: {"type":"BuildCompleted","approved_by":"cuviper","base_ref":"beta","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8477186215/job/23229779144"},"merge_sha":"c1195518c8ad7f37d15ecb82f39dec3da97f9c2f"} -->
